block/stateupdate_validator.go

```diff
@@ -0,0 +1,162 @@
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/dymensionxyz/dymint/da"
+	"github.com/dymensionxyz/dymint/settlement"
+	"github.com/dymensionxyz/dymint/types"
+)
+
+// StateUpdateValidator is a validator for messages gossiped in the p2p network.
+type StateUpdateValidator struct {
+	logger       types.Logger
+	blockManager *Manager
+}
+
+// NewValidator creates a new Validator.
+func NewStateUpdateValidator(logger types.Logger, blockManager *Manager) *StateUpdateValidator {
+	return &StateUpdateValidator{
+		logger:       logger,
+		blockManager: blockManager,
+	}
+}
+
+// ValidateStateUpdate validates that the blocks from the state info are available in DA,
+// that the information included in the Hub state info matches the blocks retrieved from DA
+// and those blocks are the same that are obtained via P2P.
+func (v *StateUpdateValidator) ValidateStateUpdate(batch *settlement.ResultRetrieveBatch) error {
+	v.logger.Debug("validating state update", "start height", batch.StartHeight, "end height", batch.EndHeight)
+
+	// validate that the DRS version specified batch matches the node state
+	err := v.validateDRS(batch.StartHeight, batch.EndHeight)
+	if err != nil {
+		return err
+	}
+
+	// load blocks applied from P2P for the batch heights
+	p2pBlocks := make(map[uint64]*types.Block)
+	for height := batch.StartHeight; height <= batch.EndHeight; height++ {
+		source, err := v.blockManager.Store.LoadBlockSource(height)
+		if err != nil {
+			continue
+		}
+		block, err := v.blockManager.Store.LoadBlock(height)
+		if err != nil {
+			continue
+		}
+		if source == types.Gossiped.String() || source == types.BlockSync.String() {
+			p2pBlocks[block.Header.Height] = block
+		}
+	}
+
+	// load all DA blocks from the batch to be validated
+	daBlocks := []*types.Block{}
+	var daBatch da.ResultRetrieveBatch
+	for {
+		daBatch = v.blockManager.Retriever.RetrieveBatches(batch.MetaData.DA)
+		if daBatch.Code == da.StatusSuccess {
+			break
+		}
+	}
+	for _, batch := range daBatch.Batches {
+		daBlocks = append(daBlocks, batch.Blocks...)
+	}
+
+	// validate DA blocks against the state update
+	err = v.ValidateDaBlocks(batch, daBlocks)
+	if err != nil {
+		return err
+	}
+
+	// validate P2P blocks against DA blocks
+	err = v.ValidateP2PBlocks(daBlocks, p2pBlocks)
+	if err != nil {
+		return err
+	}
+
+	// update the last validated height to the batch last block height
+	v.blockManager.UpdateLastValidatedHeight(batch.EndHeight)
+
+	return nil
+}
+
+// ValidateP2PBlocks basically compares that the blocks applied from P2P are the same blocks included in the batch and retrieved from DA.
+// Since DA blocks have been already validated against Hub state info block descriptors, if P2P blocks match with DA blocks, it means they are also validated against state info block descriptors.
+func (v *StateUpdateValidator) ValidateP2PBlocks(daBlocks []*types.Block, p2pBlocks map[uint64]*types.Block) error {
+	// nothing to compare
+	if len(p2pBlocks) == 0 {
+		return nil
+	}
```

Comment:
can you remove it? it's unnecessary micro optimization

Reply:
i dont think is that micro. i moved outside the function as suggested by omri.